### DISCUSSION TITLE
How to query for domain of a property

### DISCUSSION THREAD

**Discussion by AbhaMoitra** [SemanticApplicationDesignLanguage/sadl#897] [Q&A] [answered]
In the attached project, property analysisConfiguration is set in file ANALYSIS.sadl and then its domain is extended in SRI-OverlayCheckIssue.sadl. How do I write a query to get this extended domain. (More comments in the files and my query is in OverlayChecks.sadl.) 

Also, there is a type checking warning in SRI-OverlayCheckIssue.sadl that I think should not be present (more details in comments in the file and older versions of SADL does not have a type checking warning for the same file).

My version and settings:
Using SADL version 3.5.0.202202111752; using Augmented Reasoner and
set it so that "Property range specification required" is turned off 
and selected "Type checking issues as warning only"
and turned off "Translate multiple-class domain or range as union"
[OverlayCheckIssue.zip](https://github.com/SemanticApplicationDesignLanguage/sadl/files/8170532/OverlayCheckIssue.zip)

**Comment by crapo** [comment]
Here's a more complete description of what is defined in the OverLayCheckIssue project.

1. Property _used_ is defined in **PROV-S.sadl** as an _owl:ObjectProperty_ with domain _ACTIVITY_ and range _ENTITY_.
2. Property _analysisConfiguration_ is defined in **ANALYSIS.sadl** as a sub-property of _used_. (Line 41 is identified as the definition.) Line 40 adds an _rdfs:comment_ annotation and specifies the more restricted domain _ANALYSIS_, which is a subclass of the domain of _used_, and specifies the same range _ENTITY_.
3. Lines 13 and 18 of **SRI-OverlayCheckIssue.sadl** add _owl:AllValuesFrom_ restrictions on property _analysisConfiguration_ for the two different classes identified in the two statements. These are fine.
4. Line 21 of **SRI-OverlayCheckIssue.sadl** adds an _owl:AllValuesFrom_ restriction on property _analyzedWith_ for class _DaikonInvariantDetection_. This is also fine.
5. Line 22 tries to re-specify the domain of _analyzedWith_, which is defined in **ANALYSIS.sadl** as a sub-property of _wasAssociatedWith_ with domain _ANALYSIS_ and range _AGENT_. This causes a warning, as expected. Since the domain of _analyzedWith_ is already specified as _ANALYSIS_, to now say that the domain of _analyzedWith_ is _DaikonInvariantDetection_, a sub-class of _ACTIVITY_, is to say that the domain of _analyzedWith_ is the union of the classes _ACTIVITY_ and _DaikonInvariantDetection_, but that union is just the class _ACTIVITY_. Hence the warning. Note that a property domain is NOT a restriction. What are you trying to say with line 22? That only instances of _DaikonInvariantDetection_ can be the subject of triples with _analyzedWith_ as predicate? But you've already said that the domain is _ACTIVITY_...

**Reply by AbhaMoitra** [reply]
@crapo : thank you for the detailed reply. Line 22 in SRI-OverlayCheckIssue.sadl is putting a cardinality constraint. Is there a way to re-write the last 3 lines of this file to avoid a warning? I remember at an earlier point that combining range and cardinality into a single statement did not translate correctly so we split out the range and cardinality constraint but maybe it is ok now?.

**Reply by crapo** [reply]
@AbhaMoitra , if all you want to do is place a cardinality constraint on property _analysedWith_ on class _DaikonInvariantDetection_, you already have the correct SADL syntax in [commented out] line 23: "analyzedWith of DaikonInvariantDetection has exactly 1 value." Hopefully now the answer to your question on line 25, "why a warning for the line 22 but not for line 23" is clear--line 22 tries to re-specify the domain as well as add a cardinality constraint whereas line 23 just adds a cardinality constraint.

**Reply by AbhaMoitra** [reply]
@crapo : of course, thank you.

**Comment by crapo** [comment, marked as answer]
As far as the query goes. you can find all properties with their domain and range with this query:
`select ?d ?p ?r where {?p <rdfs:domain> ?d . ?p <rdfs:range> ?r}`.

You can get the domain and range of analysisConfiguration only with this query:
`select ?d ?r where {<analysisConfiguration> <rdfs:domain> ?d . <analysisConfiguration> <rdfs:range> ?r}`.

Note that the second query will give results

```
     "d","r"
     "http://arcos.rack/ANALYSIS#ANALYSIS","http://arcos.rack/PROV-S#ENTITY"
```
because line 22 doesn't add anything to the OWL file, for the reason explained above. Hence only the originally specified domain and range are found.
